Snippet updates for Response Caching Middleware /3 (#24444)
dotnet/AspNetCore.Docs: aspnetcore/performance/caching/middleware.md

monikerRange: '>= aspnetcore-3.1'
ms.author: riande
ms.custom: mvc
ms.date: 1/1/2022
no-loc: [Home, Privacy, Kestrel, appsettings.json, "ASP.NET Core Identity", cookie, Cookie, Blazor, "Blazor Server", "Blazor WebAssembly", "Identity", "Let's Encrypt", Razor, SignalR]
uid: performance/caching/middleware
---
# Response Caching Middleware in ASP.NET Core

By [John Luo](https://github.com/JunTaoLuo) and [Rick Anderson](https://twitter.com/RickAndMSFT)

::: moniker range=">= aspnetcore-6.0"

This article explains how to configure [Response Caching Middleware](https://github.com/dotnet/aspnetcore/blob/main/src/Middleware/ResponseCaching/src/ResponseCachingMiddleware.cs) in an ASP.NET Core app. The middleware determines when responses are cacheable, stores responses, and serves responses from cache. For an introduction to HTTP caching and the [`[ResponseCache]`](xref:Microsoft.AspNetCore.Mvc.ResponseCacheAttribute) attribute, see [Response Caching](xref:performance/caching/response).

The Response caching middleware:

* Implements the standard HTTP caching semantics, that is it caches based on HTTP cache headers like proxies do.

> * Is typically not beneficial for UI apps such as Razor Pages. Output caching is being considered for the next version of ASP.NET Core, which will benefit UI apps. With output caching, the app, not the browser, decides what should be cached. For more information, see [this GitHub issue](https://github.com/dotnet/aspnetcore/issues/27387).

Author: ```suggestion
* Is typically not beneficial for UI apps such as Razor Pages. Output caching is being considered for the next version of ASP.NET Core, which will benefit UI apps. With output caching, configuration decides what should be cached independently of HTTP headers. For more information, see [this GitHub issue](https://github.com/dotnet/aspnetcore/issues/27387).
```

> * May be beneficial for public GET or HEAD API requests from clients where the [Conditions for caching](#cfc) are met.

Author: @adityamandaleeka @halter73 please review these two lines. Feel free to look at the rest of the doc.

Reviewer: > With output caching, the app, not the browser, decides what should be cached.

It would imply that this is the browser that decides what should be cached. Though even with Response Caching this is not the case, it's "response headers" based. I would suggest:

"With output caching, configuration decides what should be cached independently of HTTP headers."

<!--Postman:  GET: Headers > Postman > go to settings > Send no-cache header -->
## Configuration

> In `Program.cs`, add the Response Caching Middleware to the service collection and configure the app to use the middleware with the <xref:Microsoft.AspNetCore.Builder.ResponseCachingExtensions.UseResponseCaching*> extension method. `UseResponseCaching` adds the middleware to the request processing pipeline:

Reviewer 2: ```suggestion
In `Program.cs`, add the Response Caching Middleware services to the service collection and configure the app to use the middleware with the <xref:Microsoft.AspNetCore.Builder.ResponseCachingExtensions.UseResponseCaching*> extension method. `UseResponseCaching` adds the middleware to the request processing pipeline:
```

> [!WARNING]
> Responses containing content for authenticated clients must be marked as not cacheable to prevent the middleware from storing and serving those responses. See [Conditions for caching](#conditions-for-caching) for details on how the middleware determines if a response is cacheable.

> The preceding code typically won't return a cached value to a browser. Use [Fiddler](https://www.telerik.com/fiddler), [Postman](https://www.getpostman.com/), or other tool that can explicitly set request headers and are preferred for testing caching. For more information, see [Troubleshooting](#troubleshooting) in this article.

Reviewer 2: ```suggestion
The preceding code typically doesn't return a cached value to a browser. Use [Fiddler](https://www.telerik.com/fiddler), [Postman](https://www.getpostman.com/), or another tool that can explicitly set request headers and is preferred for testing caching. For more information, see [Troubleshooting](#troubleshooting) in this article.
```

## Options

> Response caching options are shown in the following table.

Reviewer 3: I just noticed the size limit on the cache is 100MB. I feel like we might be hitting that, but how can I find out? Does the middleware log warnings?

Reviewer 3: "Overcapacity compaction triggered" is logged at Debug level by `Microsoft.Extensions.Caching.Memory.MemoryCache`. [Source](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L359).

Reviewer 3: This comment thread can be resolved @Rick-Anderson, unless you think it's worth mentioning that Response Caching uses IMemoryCache, update docs on Memory Caching to say this log message is written, and then refer to that from here as a way to know if you're hitting limits. That could have value; would have saved me a few mins.

Reviewer 3: Turns out the response caching middleware spawns its own `MemoryCache` instance but doesn't pass it a logger so it gets the null logger and you don't get anything coming out. At least I think that's what's happening. 

https://github.com/dotnet/aspnetcore/blob/c85baf8db0c72ae8e68643029d514b2e737c9fae/src/Middleware/ResponseCaching/src/ResponseCachingMiddleware.cs#L39-L54

Author: > Turns out the response caching middleware spawns its own `MemoryCache` instance but doesn't pass it a logger factory so it gets a no-op logger and you don't get anything coming out. At least I think that's what's happening.
> 
> https://github.com/dotnet/aspnetcore/blob/c85baf8db0c72ae8e68643029d514b2e737c9fae/src/Middleware/ResponseCaching/src/ResponseCachingMiddleware.cs#L39-L54

@benaadams is that true?

Reviewer 3: In the interest of not hijacking this PR we can discuss this on the [issue I created in aspnetcore](https://github.com/dotnet/aspnetcore/issues/39300). I've proposed a change and if that goes through I'll make sure to open a new PR to document the logging.
